submit/workflow/processor.py

```diff
@@ -0,0 +1,73 @@
+"""Defines submission stages and workflows supported by this UI."""
+
+from typing import Optional, Dict
+
+from arxiv.submission.domain import Submission
+from dataclasses import field, dataclass
+from . import WorkflowDefinition, Stage
+
+
+@dataclass
+class WorkflowProcessor:
+    """Class to handle a submission moving through a WorkflowDefinition.
+
+    The seen methods is_seen and mark_seen are handled with a Dict. This class
+    doesn't handle loading or saving that data.
+    """
+    workflow: WorkflowDefinition
+    submission: Submission
+    seen: Dict[str, bool] = field(default_factory=dict)
+
+    def is_complete(self) -> bool:
+        """Determine whether this workflow is complete."""
+        return bool(self.submission.is_finalized)
+
+    def next_stage(self, stage: Optional[Stage]) -> Optional[Stage]:
+        """Get the stage after the one in the parameter."""
+        return self.workflow.next_stage(stage)
+    
+    def can_proceed_to(self, stage: Optional[Stage]) -> bool:
+        """Determine whether the user can proceed to a stage."""
+        if stage is None:
+            return True
+        must_be_done = self.workflow.order if stage == self.workflow.confirmation \
+            else self.workflow.iter_prior(stage)
+        return all(map(self.is_done, must_be_done))
+
+    def current_stage(self) -> Optional[Stage]:
```

Comment:
current_stage() gets the first stage that is not yet done. It might need a different name? first_incomplete_stage()?

The current code does force the user to go one step at a time through the workflow. It allows skipping optional stages.